File: day3/part1.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int main() {
	freopen("input.txt", "r", stdin);
	int num_instructions = 0;
	string binary_string;
	cin >> binary_string;
	int num_bits = binary_string.length();
	vector<int> frequencies(num_bits);
	int data_length = 0;
	// Calculate frequency of '1' bits at each position
	while (cin) {
		for (int i = 0; i < binary_string.length(); i++) {
			if (binary_string[i] == '1') {
				frequencies[i]++;
			}
		}
		cin >> binary_string;
		data_length++;
	}
	// Calculate gamma rate based on frequencies
	int gamma_rate = 0;
	for (int frequency : frequencies) {
		gamma_rate <<= 1;
		if (frequency > data_length/2) {
			gamma_rate++;
		}
	}
	// Calculate epsilon rate based on gamma rate (binary inverse)
	int epsilon_rate = (1 << num_bits) - 1 - gamma_rate;

	int power_consumption = gamma_rate * epsilon_rate;
	cout << "Result: " << power_consumption << '\n';
}

File: day3/part1.py
# Read file input
with open("input.txt", "r") as f:
	data = [str(i).rstrip('\n') for i in f]

num_bits = len(data[0])

# Calculate frequency of '1' bits at each position
frequencies = [0]*num_bits
for binary_string in data:
	for i,c in enumerate(binary_string):
		if c == '1':
			frequencies[i] += 1
# Calculate gamma rate based on frequencies
gamma_rate = 0
for frequency in frequencies:
	gamma_rate <<= 1
	if frequency > len(data)//2:
		gamma_rate += 1
# Calculate epsilon rate based on gamma rate (binary inverse)
epsilon_rate = 2**num_bits - 1 - gamma_rate

power_consumption = gamma_rate * epsilon_rate
print(f"Result: {power_consumption}")
